pytools/graph.py

```diff
@@ -394,4 +395,28 @@ def compute_induced_subgraph(graph: Mapping[T, Set[T]],
 
 # }}}
 
+
+# {{{ validate graph
+
+def validate_graph(graph: Mapping[T, Collection[T]]) -> None:
+    """
+    Validates that all successor nodes of each node in *graph* are keys in
+    *graph* itself. Raises a :class:`ValueError` if not.
+
+    :arg graph: A :class:`collections.abc.Mapping` representing a directed
+        graph. The dictionary contains one key representing each node in the
+        graph, and this key maps to a :class:`collections.abc.Collection` of nodes
+        that are connected to the node by outgoing edges.
+    """
+    seen_nodes: Set[T] = set()
+
+    for children in graph.values():
+        seen_nodes.update(children)
+
+    if not seen_nodes <= graph.keys():
+        raise ValueError(
+            f"invalid graph: {graph.keys()}, seen nodes: {seen_nodes}")
```

Comment:
This should probably mention the offending key. Also, I think printing the entire data structure is not a good idea: this could be very large.

Reply:
Good idea - what do you think of 1aa6181?

Reply:
LGTM